# 1.Context
The team is acting in the role of Producer. Our goal is to develop a release strategy for anime that maximizes the chance of success. The team wants to leverage the current dataset to extract useful insights and build a machine learning model to quantify the impact of various factors on the success of an anime title.

In [63]:
from pathlib import Path
import pandas as pd
import numpy as np
from rapidfuzz import fuzz
import pandera as pa
import plotly.express as px
import seaborn as sns
import os
import plotly.io as pio
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import re
import ast


In [2]:
import kagglehub
# Download latest version
path = kagglehub.dataset_download("dbdmobile/myanimelist-dataset")

print("Path to dataset files:", path)

# List all files in the dataset directory
folder = os.listdir(path)
print("Files in dataset directory:", folder)
for file in folder:
    print(file)

c:\Users\ChooChoo\AppData\Local\Programs\Python\Python312\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Path to dataset files: C:\Users\ChooChoo\.cache\kagglehub\datasets\dbdmobile\myanimelist-dataset\versions\5
Files in dataset directory: ['anime-dataset-2023.csv', 'anime-filtered.csv', 'final_animedataset.csv', 'user-filtered.csv', 'users-details-2023.csv', 'users-score-2023.csv']
anime-dataset-2023.csv
anime-filtered.csv
final_animedataset.csv
user-filtered.csv
users-details-2023.csv
users-score-2023.csv


In [3]:
current_dir = Path.cwd()

project_root = current_dir.parent 

raw_data_path = project_root / "data" / "raw" / "anime-dataset-2023.csv"
processed_data_path = project_root / "data" / "processed" / "prepared_data.csv"

df_anime_dataset_2023 = pd.read_csv(raw_data_path)
df_anime_dataset_2023_prep = pd.read_csv(processed_data_path)

In [4]:
# # Tạo dataframe từ dữ liệu gốc
# csv_file = f"{path}/anime-dataset-2023.csv"
# df_anime_dataset_2023 = pd.read_csv(csv_file)

# #Tạo dataframe từ dữ liệu đã qua chỉnh sửa
# csv_file2 = r"C:\Users\ChooChoo\Downloads\prepared_data.csv"  #Thay đường dẫn ở đây
# df_anime_dataset_2023_prep = pd.read_csv(csv_file2)

In [5]:
pio.templates.default = "plotly_white"

In [6]:
print(df_anime_dataset_2023.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24905 entries, 0 to 24904
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   anime_id      24905 non-null  int64 
 1   Name          24905 non-null  object
 2   English name  24905 non-null  object
 3   Other name    24905 non-null  object
 4   Score         24905 non-null  object
 5   Genres        24905 non-null  object
 6   Synopsis      24905 non-null  object
 7   Type          24905 non-null  object
 8   Episodes      24905 non-null  object
 9   Aired         24905 non-null  object
 10  Premiered     24905 non-null  object
 11  Status        24905 non-null  object
 12  Producers     24905 non-null  object
 13  Licensors     24905 non-null  object
 14  Studios       24905 non-null  object
 15  Source        24905 non-null  object
 16  Duration      24905 non-null  object
 17  Rating        24905 non-null  object
 18  Rank          24905 non-null  object
 19  Popu

# 2.Data Frame Introduction

1. Thông tin cơ bản và nhận dạng (Basic Identification & Description)
- **anime_id**: ID duy nhất cho mỗi anime.
- **Name**: Tên của anime bằng ngôn ngữ gốc.
- **English name**: Tên tiếng Anh của anime.
- **Other name**: Tên bản địa hoặc tựa đề của anime.
- **Synopsis**: Mô tả hoặc tóm tắt ngắn gọn về cốt truyện của anime.
- **Genres**: Các thể loại của anime, được phân tách bằng dấu phẩy.
- **Image URL**: URL của hình ảnh hoặc poster của anime.

2. Chi tiết sản xuất và kỹ thuật (Production & Technical Details)
- **Type**: Loại anime.
- **Source**: Vật liệu gốc của anime.
- **Producers**: Các công ty sản xuất hoặc nhà sản xuất của anime.
- **Studios**: Các studio hoạt hình đã thực hiện anime.
- **Licensors**: Các nhà cấp phép của anime.
- **Episodes**: Số lượng tập trong anime.
- **Duration**: Thời lượng của mỗi tập phim.

3. Thông tin phát sóng và phát hành (Release & Airing Information)
- **Aired**: Ngày anime được phát sóng.
- **Premiered**: Mùa và năm anime ra mắt.
- **Status**: Trạng thái của anime.

4. Chỉ số tương tác người xem và hiệu suất (Audience Engagement & Performance Metrics)
- **Score**: Điểm được trao cho anime.
- **Rating**: Xếp hạng độ tuổi của anime.
- **Rank**: Xếp hạng của anime dựa trên mức độ phổ biến hoặc các tiêu chí khác.
- **Popularity**: Xếp hạng mức độ phổ biến của anime.
- **Favorites**: Số lần anime được người dùng đánh dấu là yêu thích.
- **Scored By**: Số lượng người dùng đã chấm điểm anime.
- **Members**: Số lượng thành viên đã thêm anime vào danh sách của họ trên nền tảng.

*Unnecessary Feature*
- Tiến hành drop feature không cần thiết khỏi dataframe

In [7]:
df_anime_dataset_2023 = df_anime_dataset_2023.drop(columns=['English name', 'Other name', 'Synopsis', 'Image URL'])

# 3.Data Insight

## 3.6.Insight Question 6
The **Source** feature is invaluable for understanding the origins of anime, revealing whether a title is adapted from manga, light novels, games, or is an entirely original work. This information can shed light on industry trends, adaptation strategies, and even predict potential fan bases or production approaches. Our initial analysis will examine the raw 'Source' column to identify the most prevalent origins of anime within our dataset. By intentionally bypassing any preprocessing, we aim to uncover the immediate distribution of these source categories and, crucially, to identify if any non-standard entries or placeholder values are present, which could distort our understanding of the true source landscape and necessitate data cleaning.

### 3.6.1.Misleading Insight
- What is the frequency of each unique source category present in the raw **Source** column of the dataset?

In [16]:
# 1. Thiết lập theme mặc định là nền trắng (Học từ Code 2)
pio.templates.default = "plotly_white"

# 2. Xử lý dữ liệu
source_counts = df_anime_dataset_2023['Source'].value_counts().reset_index()
source_counts.columns = ['Source', 'Count']

# Sắp xếp ascending=True để khi vẽ ngang, thanh lớn nhất sẽ nằm trên cùng (Declutter)
source_counts = source_counts.sort_values('Count', ascending=True)

# 3. Tạo logic màu sắc (Học từ Code 2)
# Nếu là 'Unknown' thì màu xám, còn lại là màu Tím (#8B5CF6)
base_color = "#8B5CF6"
unknown_color = "#B0B0B0"

# Tạo list màu tương ứng với thứ tự dữ liệu đã sắp xếp
colors = [
    unknown_color if str(s).strip().upper() == "UNKNOWN" else base_color
    for s in source_counts["Source"]
]

# 4. Tạo biểu đồ Horizontal Bar
fig = px.bar(
    source_counts,
    x='Count',      # Số lượng nằm ngang
    y='Source',     # Tên nguồn nằm dọc
    orientation='h', # Chuyển sang dạng thanh ngang
    title='Frequency of Anime Sources (Raw Data)',
    labels={'Source': 'Source Type', 'Count': 'Number of Anime'},
    text='Count'
)

# 5. Tinh chỉnh giao diện (Kết hợp Code 1 và Code 2)
fig.update_traces(
    marker_color=colors,           # Áp dụng logic màu đã tạo
    marker_line_color='white',
    marker_line_width=1.2,
    textposition='outside',
    texttemplate='%{text:,}',
    hovertemplate='<b>Source:</b> %{y}<br><b>Count:</b> %{x:,}<extra></extra>'
)

# 6. Tinh chỉnh Layout (Modern & Clean)
fig.update_layout(
    title={
        'text': 'Frequency of Anime Sources (Raw Data)',
        'x': 0.5,
        'xanchor': 'center',
        'font': dict(size=22, family='Inter', weight='bold', color='#1F2937')
    },
    # Trục X: Số lượng
    xaxis=dict(
        title='Number of Anime',
        title_font=dict(size=14, color='#4B5563', family='Inter'),
        showgrid=False,        # Giữ grid dọc mờ để dễ so sánh độ dài
        gridcolor='#F3F4F6',  # Grid màu rất nhạt
        showline=False,       # Bỏ đường kẻ trục đậm
        tickformat=',',
        zeroline=False
    ),
    # Trục Y: Tên nguồn (Nơi cần ticksuffix)
    yaxis=dict(
        title='',             # Bỏ tiêu đề trục Y vì tên Source đã rõ ràng
        tickfont=dict(size=13, color='#4B5563', family='Inter'),
        ticksuffix="  ",      # <--- Thêm khoảng cách theo yêu cầu của bạn
        showgrid=False,
        showline=False,       # Bỏ đường kẻ trục cho thoáng
        zeroline=False
    ),
    bargap=0.25,
    height=600,
    margin=dict(t=80, l=120, r=40, b=80),
    hoverlabel=dict(
        bgcolor='white',
        bordercolor='#8B5CF6',
        font=dict(size=13, family='Inter')
    )
)

fig.show()

**Conflict**
The initial insight derived from the **Source** feature (showing the frequency of anime origins) is skewed and misleading due to issues related to its data representation. The raw state of this column prevents an accurate understanding of the true prevalence of each source.
- Error 1: Mismatched Data Type - **Source** as a Generic Object (String) Type.
  - Impact on Insight (Bar/Pie Chart): While Plotly can render categories from an object (string) column, the generic nature of this data type means there's no inherent enforcement of valid source categories. This allows for the inclusion of arbitrary string values, which are then treated as legitimate categories by the plotting library. This lack of type specificity contributes to the problem of non-standard entries being interpreted as meaningful data, rather than flagging them as potential errors or missing information.
- Error 2: Non-standard Missing Value Representation - 'UNKNOWN' instead of NaN.
  - Impact on Insight (Bar/Pie Chart): The presence of the string 'UNKNOWN' within the **Source** column is a significant distortion. Instead of being recognized as a missing value (which Plotly would typically ignore or handle gracefully), 'UNKNOWN' is treated as a distinct and valid source category. Consequently, in both a bar chart and a pie chart, 'UNKNOWN' will appear as a prominent category, often ranking highly in frequency. This falsely inflates its importance and misrepresents the true distribution of actual anime sources, making it difficult to discern the genuine origins of anime in the dataset.


### 3.6.2.Correct Insight

In [17]:
# 1. Giữ theme mặc định
pio.templates.default = "plotly_white"

# 2. Xử lý dữ liệu (Cleaned Data)
# Đếm tần suất
source_counts = df_anime_dataset_2023_prep['Source'].value_counts().reset_index()
source_counts.columns = ['Source', 'Count']

# Sắp xếp để thanh dài nhất nằm trên cùng
source_counts = source_counts.sort_values('Count', ascending=True)

# 3. Tạo biểu đồ Horizontal Bar
fig = px.bar(
    source_counts,
    x='Count',
    y='Source',
    orientation='h',
    title='Frequency of Anime Sources (Cleaned Data)', 
    labels={'Source': 'Source Type', 'Count': 'Number of Anime'},
    text='Count',
    color='Count',
    color_continuous_scale=['#DDD6FE', '#8B5CF6', '#4C1D95']
)

# 4. Tinh chỉnh giao diện Traces (Đồng nhất style)
fig.update_traces(
    marker_line_color='white',
    marker_line_width=1.2,
    textposition='outside',
    texttemplate='%{text:,}',
    hovertemplate='<b>Source:</b> %{y}<br><b>Count:</b> %{x:,}<extra></extra>'
)

# 5. Tinh chỉnh Layout (Copy style từ biểu đồ Raw để đồng bộ)
fig.update_layout(
    title={
        'text': 'Frequency of Anime Sources (Cleaned Data)',
        'x': 0.5,
        'xanchor': 'center',
        'font': dict(size=22, family='Inter', weight='bold', color='#1F2937')
    },
    # TRỤC X: Hiển thị Grid, Format số
    xaxis=dict(
        title='Number of Anime',
        title_font=dict(size=14, color='#4B5563', family='Inter'),
        tickfont=dict(size=11, color='#6B7280'),
        showgrid=False,       
        gridcolor='#F3F4F6',  # Grid màu nhạt
        showline=False,       # Bỏ đường kẻ trục đậm
        zeroline=False,
        tickformat=','        # Dấu phẩy phân cách hàng nghìn
    ),
    # TRỤC Y: Chứa nhãn Source, Ticksuffix
    yaxis=dict(
        title='',             # Bỏ tiêu đề trục "Source" cho thoáng
        tickfont=dict(size=13, color='#4B5563', family='Inter'),
        ticksuffix="  ",      # <--- QUAN TRỌNG: Tạo khoảng cách với thanh bar
        showgrid=False,       # Tắt grid ngang
        showline=False,
        zeroline=False
    ),
    coloraxis_showscale=False,
    plot_bgcolor='white',
    paper_bgcolor='white',    # Đồng bộ màu nền
    bargap=0.25,              # Khoảng cách giữa các thanh
    height=600,
    margin=dict(t=80, l=120, r=40, b=80), # Lề trái rộng (120) cho tên Source dài
    hoverlabel=dict(
        bgcolor='white',
        bordercolor='#8B5CF6',
        font=dict(size=13, family='Inter', color='#1F2937')
    )
)

fig.show()

##### Initial Analysis
After removing the **Unknown** category, the distribution of anime by Source has become more reliable and reshaped the market landscape. The Anime industry show a strong reliance on the traditional content sources with the **Original creations** dominating the largest share, followed by the **Manga** at a significant distance. 

Game-based and Visual Novel adaptations occupy the next major tiers, while sources such as Light Novels, Web Manga, Music, and niche formats like 4-koma Manga or Picture Books contribute much smaller proportions.

$\implies$ The Anime industry tends to produce a high number of Original works to avoid dependence on licensing, or relies on the safety of existing popular Manga. ***However, does “a large quantity” necessarily mean “good quality”?***


**Deep Insight**:
- Điểm số trung bình của mỗi loại Source là bao nhiêu?
- Các type anime như Movie và Serie thường chuyển thể từ các Source nào.

In [ ]:
import plotly.express as px
import pandas as pd

# --- 1. DATA PROCESSING ---
avg_score_by_source = df_anime_dataset_2023_prep.groupby('Source', dropna=True)['Score'].mean().reset_index()
avg_score_by_source = avg_score_by_source.dropna(subset=['Score']).sort_values('Score', ascending=False)

# --- 2. PLOTTING ---
fig_avg = px.bar(
    avg_score_by_source,
    x='Score',
    y='Source',
    orientation='h',
    text='Score',
    color='Score', 
    color_continuous_scale=[ '#DDD6FE', '#4C1D95' ],
)

# --- 3. STYLING TRACES ---
fig_avg.update_traces(
    marker_line_width=0, 
    textposition='outside',
    texttemplate='%{text:.2f}',
    hovertemplate='<b>Source:</b> %{y}<br><b>Avg Score:</b> %{x:.2f}<extra></extra>'
)

# --- 4. LAYOUT & AXES ---
fig_avg.update_layout(
    title={
        'text': '<b>Average Score by Source</b>', 
        'y': 0.95, 'x': 0.5, 'xanchor': 'center', 'yanchor': 'top'
    },
    font=dict(family='inter', size=13, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    height=600,
    margin=dict(t=80, l=150, r=50, b=50), 
    showlegend=False,
    coloraxis_showscale=False,

    # Trục X: Ẩn hoàn toàn
    xaxis=dict(visible=False),

    # Trục Y: Tăng cỡ chữ lên 14
    yaxis=dict(
        title=None,
        showgrid=False,
        showline=False,
        ticklen=0,
        autorange="reversed",
        ticksuffix="  ",
        tickfont=dict(size=14) 
    )
)

fig_avg.show()

#### Deep Analysis
##### **A. The Paradox of "Quantity" vs. "Quality"** 
*   **The Fall of "Original":** Despite ranking **#1** in production release volume, the average score for *Original* works falls into the low tier (**6.07**).
    *   **Reason:** Producing an original script is immensely risky. Without an existing fanbase, the storyline has not yet been tested in the market. The data suggests a **large number of low-quality or short-lived original anime being mass-produced**, which lowers the overall average score.
*   **The Rise of the "Novel" Category:**
    *   **Web Novels (7.00)** and **Light Novels (6.96)** lead the rankings in score, despite their modest production numbers.
    *   **Manga (6.83)** maintains stable performance: High volume combined with high scores.
    *   **Reason - The "Selection Bias" Effect:** A Web Novel or Light Novel typically needs to reach high popularity and demonstrate strong narrative quality before it is selected for anime adaptation. As a result, works chosen from this source category tend to have a higher probability of critical success (e.g., *Re:Zero, Sword Art Online*).

$\implies$ **Strategy for the Producer:** For safety and guaranteed high ratings, prioritize adaptations from famous **Web Novels, Light Novels, or Manga**. If choosing to produce an **Original**, accept high risks and invest heavily in the scriptwriting team to avoid the "low-score trap."

In [ ]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

# --- 1. SETUP SUBPLOTS ---
# Tạo khung chứa 1 hàng, 2 cột
fig = make_subplots(
    rows=1, cols=2,
    subplot_titles=["<b>Top Sources for TV</b>", "<b>Top Sources for Movie</b>"],
    horizontal_spacing=0.12  # Khoảng cách giữa 2 biểu đồ
)

types_of_interest = ['TV', 'Movie'] # Thứ tự hiển thị: Trái -> Phải

# Định nghĩa dải màu (Light Purple -> Deep Purple)
custom_colorscale = [ '#DDD6FE', '#4C1D95' ]

# --- 2. DATA PROCESSING & PLOTTING ---
for i, t in enumerate(types_of_interest):
    # Data Processing
    df_t = df_anime_dataset_2023_prep[df_anime_dataset_2023_prep['Type'] == t]
    counts = df_t['Source'].value_counts().reset_index()
    counts.columns = ['Source', 'Count']
    
    # Lấy Top 10 và sắp xếp giảm dần (để hiển thị đúng thứ tự từ trên xuống)
    counts = counts.head(10).sort_values('Count', ascending=False)
    
    # Tạo Trace (Bar Chart)
    fig.add_trace(
        go.Bar(
            x=counts['Count'],
            y=counts['Source'],
            orientation='h',
            text=counts['Count'],
            texttemplate='%{text:,}', # Format số có dấu phẩy
            textposition='outside',
            hovertemplate='<b>Source:</b> %{y}<br><b>Count:</b> %{x:,}<extra></extra>',
            
            # Color Logic: Map màu theo số lượng
            marker=dict(
                color=counts['Count'],
                colorscale=custom_colorscale,
                line_width=0
            ),
            name=t # Tên cho legend (nhưng ta sẽ ẩn legend)
        ),
        row=1, col=i+1 # Đặt vào cột tương ứng (1 hoặc 2)
    )

# --- 3. GLOBAL STYLING (DECLUTTER) ---

# Cập nhật Layout chung
fig.update_layout(
    title={
        'text': '<b>Source Volume Distribution According to Type: TV vs Movie</b>',
        'y': 0.95,
        'x': 0.5,
        'xanchor': 'center',
        'yanchor': 'top'
    },
    font=dict(family='Inter', size=13, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    height=600,
    width=1400, # Độ rộng đủ cho 2 chart
    margin=dict(t=80, l=50, r=50, b=50),
    showlegend=False
)

# Tinh chỉnh trục X (Ẩn số, ẩn grid) cho cả 2 biểu đồ
fig.update_xaxes(showticklabels=False, showgrid=False, showline=False, visible=False)

# Tinh chỉnh trục Y (Giữ nhãn Source, ẩn grid, đảo ngược chiều để Top 1 nằm trên)
fig.update_yaxes(
    showgrid=False, 
    showline=False, 
    ticklen=0, 
    autorange="reversed", # Đảo ngược trục Y để giá trị cao nhất nằm trên cùng
    ticksuffix="  "      # Tạo khoảng cách nhỏ giữa chữ và thanh bar
)

# Mở rộng trục X một chút để số không bị cắt (vì textposition='outside')
# Chúng ta lặp qua từng trace để lấy max value và set range
for i, t in enumerate(types_of_interest):
    df_t = df_anime_dataset_2023_prep[df_anime_dataset_2023_prep['Type']== t]
    max_val = df_t['Source'].value_counts().max()
    # Update axes cụ thể (xaxis1, xaxis2)
    if i == 0:
        fig.update_xaxes(range=[0, max_val * 1.25], row=1, col=1)
    else:
        fig.update_xaxes(range=[0, max_val * 1.25], row=1, col=2)

fig.show()

##### **B.Format and Source Compatibility (Movie vs. TV)**

According to the distribution of Type released, TV and Movie are leading the market. Selection behavior for script sources changes drastically depending on the release format:

*   **For TV Series:**
    *   **Top 3 Structure:** **Original > Manga > Light Novel**.
    *   *Analysis:* TV Series require long-form content to sustain broadcasting for 3-6 months (12-24 episodes). **Manga** and **Light Novels**, with their episodic structures and expansive world-building, are highly suitable for this format. Light Novels breaking into the Top 3 for TV series (488 titles) proves this is a "fertile ground" for serialized content.
*   **For Movies:**
    *   **Top 3 Structure:** **Original > Manga > Other**.
    *   *The Disappearance of Light Novels:* Light Novels drop significantly in the Movie category.
    *   *Analysis:* Movies have limited runtime (90-120 minutes), demanding concise, dramatic plots with clear conclusions.
        *   **Originals** dominate Movies (1,908 titles) as screenwriters can easily tailor a complete story specifically for the 2-hour format.
        *   **Manga** movies are often spin-offs of major franchises (e.g., *Conan, Doraemon*) or short one-shots.
        *   **Light/Web Novels** are often too complex and lengthy to compress into a single film, making them less prioritized for the Movie format.

## 3.7.Insight Question 7
A Producer can be understood as "investors," and an anime is essentially an "investment project." Producers are the entities that provide funding, plan, and coordinate the anime project. Normally, producing an anime requires a huge budget. To minimize financial risk in case the anime fails and causes significant losses, multiple producers often collaborate on a single anime. Consequently, the 'Producers' feature in our dataset often lists several production companies or individuals within a single entry, separated by commas. Our initial exploration of this raw column will aim to identify the most frequently appearing combinations of producers, which can hint at common collaborations or dominant production groups. By analyzing this unprocessed, aggregated data, we intend to highlight how its current string format will impede the identification of individual producers' prevalence and necessitate significant parsing and cleaning.

### 3.7.1.Misleading Insight
- What are the most frequently appearing producer combinations or producer strings as they are originally listed in the dataset?

In [20]:
# 2. Xử lý dữ liệu (Top 15 Producers)
producer_counts = df_anime_dataset_2023['Producers'].value_counts().head(15).reset_index()
producer_counts.columns = ['Producers', 'Count']

# Sắp xếp ascending=True để khi vẽ ngang, thanh lớn nhất nằm trên cùng
producer_counts = producer_counts.sort_values('Count', ascending=True)

# 3. Logic màu sắc (Gray cho Unknown, Tím cho còn lại)
base_color = "#8B5CF6"
unknown_color = "#B0B0B0"

# Tạo list màu dựa trên tên Producer
colors = [
    unknown_color if str(p).strip().upper() == "UNKNOWN" else base_color
    for p in producer_counts['Producers']
]

# 4. Tạo biểu đồ Horizontal Bar
fig = px.bar(
    producer_counts,
    x='Count',       # Số lượng nằm ngang
    y='Producers',   # Tên nằm dọc
    orientation='h', # Xoay ngang
    title='Top 15 Producer Combinations (Raw Data)',
    labels={'Producers': 'Producer Combination', 'Count': 'Number of Anime'},
    text='Count'
)

# 5. Tinh chỉnh traces (Áp dụng màu và format)
fig.update_traces(
    marker_color=colors,  # Áp dụng list màu đã tạo ở bước 3
    marker_line_color='white',
    marker_line_width=1.2,
    textposition='outside',
    texttemplate='%{text:,}',
    hovertemplate='<b>Producers:</b> %{y}<br><b>Count:</b> %{x:,}<extra></extra>'
)

# 6. Tinh chỉnh Layout (Đồng bộ style)
fig.update_layout(
    title={
        'text': 'Top 15 Producer Combinations (Raw Data)',
        'x': 0.5,
        'xanchor': 'center',
        'font': dict(size=22, family='Inter', weight='bold', color='#1F2937')
    },
    # Trục X: Số lượng (Hiển thị Grid để dễ so sánh)
    xaxis=dict(
        title='Number of Anime',
        title_font=dict(size=14, color='#4B5563', family='Inter'),
        tickfont=dict(size=11, color='#6B7280'),
        showgrid=False,
        gridcolor='#F3F4F6',
        showline=False,
        zeroline=False,
        tickformat=','
    ),
    # Trục Y: Tên Producers (Cần ticksuffix)
    yaxis=dict(
        title='',  # Bỏ tiêu đề trục Y vì tên đã rõ nghĩa
        tickfont=dict(size=14, color='#4B5563', family='Inter'),
        ticksuffix="  ",  # Tạo khoảng cách
        showgrid=False,
        showline=False,
        zeroline=False
    ),
    plot_bgcolor='white',
    paper_bgcolor='white',
    bargap=0.25,
    height=700,  # Tăng chiều cao vì có 15 dòng
    # Margin left (l) cần lớn (250) vì tên Producer Combinations thường rất dài
    margin=dict(t=80, l=250, r=40, b=80),
    hoverlabel=dict(
        bgcolor='white',
        bordercolor='#8B5CF6',
        font=dict(size=13, family='Inter', color='#1F2937')
    )
)

fig.show()

**Conflict**
The initial insights derived from the 'Producers' feature (showing the frequency of producer combinations) are significantly flawed and misleading due to critical data integrity and structural issues. The raw state of this column prevents an accurate understanding of individual producers' contributions or true prevalence.
- Error 1: Aggregated Producer Strings (Multiple Producers in One Entry).
    - Impact on Insights (Bar Charts of Producer Combinations): The primary issue is that the 'Producers' column often contains multiple producers listed within a single string entry (e.g., "Aniplex, Lantis, MAGES."). When we count the frequency of these strings, we are tallying specific combinations of producers, not the individual producers themselves. This fundamentally obscures the true involvement and popularity of any single production company. For example, "Aniplex" might be a highly active producer, but its individual prevalence is hidden within numerous unique combinations. The bar charts will show these distinct, often lengthy, combination strings as categories, making it impossible to determine which individual producers are most active or involved across the dataset. We are gaining insight into producer teams or bundles, rather than the independent entities.
- Error 2: Non-standard Missing Value Representation - 'UNKNOWN' instead of NaN.
    - Impact on Insights (Bar Charts of Producer Combinations): The presence of the string 'UNKNOWN' for missing producer information is treated as a valid and distinct producer combination by Plotly. This means 'UNKNOWN' will likely appear as one of the most frequent "producer combinations" in the bar charts. Its inclusion as a prominent category inflates its perceived importance, distorts the true ranking of actual producer collaborations, and distracts from analyzing meaningful production data. A correctly handled missing value (standard NaN) would typically be excluded from such frequency counts by default, providing a clearer and more accurate insight into actual producer involvement.

### 3.7.2.Correct Insight

In [22]:
import plotly.express as px
import plotly.io as pio
import pandas as pd

# 1. Setup Theme
pio.templates.default = "plotly_white"

# --- 2. DATA PROCESSING (DEEP CLEANING - KEEP AS IS) ---
prod_df = df_anime_dataset_2023_prep[['Producers']].copy()
prod_df['Producers'] = prod_df['Producers'].astype(str)

# REGEX CLEANING
prod_df['Producers'] = prod_df['Producers'].str.replace(r"[\[\]\'\"]", "", regex=True)

# SPLIT & EXPLODE
prod_df['Producers'] = prod_df['Producers'].str.split(',')
prod_df = prod_df.explode('Producers')

# TRIM
prod_df['Producers'] = prod_df['Producers'].str.strip()

# FILTER
exclude_list = ["UNKNOWN", "NONE", "NAN", "NULL"]
prod_df = prod_df[~prod_df['Producers'].str.upper().isin(exclude_list)]
prod_df = prod_df[prod_df['Producers'] != ""]

# --- 3. AGGREGATION ---
top_producers = prod_df['Producers'].value_counts().head(20).reset_index()
top_producers.columns = ['Producers', 'Count']
top_producers = top_producers.sort_values('Count', ascending=True)

# --- 4. VISUALIZATION ---
fig = px.bar(
    top_producers,
    x='Count',
    y='Producers',
    orientation='h',
    text='Count',
    title='<b>Market Leaders: Top 20 Most Active Individual Producers</b>',
    color='Count',
    color_continuous_scale=[ '#DDD6FE', '#4C1D95' ]
)

# Tinh chỉnh Traces
fig.update_traces(
    texttemplate='%{text:,}',
    textposition='outside',
    hovertemplate='<b>Producer:</b> %{y}<br><b>Projects:</b> %{x:,}<extra></extra>'
)

# --- 5. LAYOUT ---
fig.update_layout(
    title={
        'text': '<b>Top 20 Most Active Individual Producers</b>',
        'y': 0.95,
        'x': 0.5,
        'xanchor': 'center',
        'yanchor': 'top'
    },
    font=dict(family='Inter', size=13, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    height=750, 
    coloraxis_showscale=False,
    margin=dict(t=80, l=150, r=50, b=80),

    # --- CẤU HÌNH TRỤC X (Số lượng) ---
    xaxis=dict(
        title="Number of Anime Produced",
        title_font=dict(size=14, color='#4B5563', family='Inter'),
        showticklabels=True, 
        tickfont=dict(size=14, color='#6B7280'),
        
        # Grid settings
        showgrid=False,        
        gridcolor='#F3F4F6',
        zeroline=False,
        
        # --- QUAN TRỌNG: CHỈNH BƯỚC NHẢY 200 ---
        tick0=0,     
        dtick=200,   
        tickformat=',' 
    ),

    # --- CẤU HÌNH TRỤC Y (Tên Producers) ---
    yaxis=dict(
        title="",
        tickfont=dict(size=14, color='#333333', family='Inter'),
        ticksuffix="  ",    # Khoảng cách đẹp
        showgrid=False,
        showline=False,
        zeroline=False
    )
)

fig.show()

#### Initial insight



**Deep Insight**

- Compare the average Score between anime with `Producers = "UNKNOWN"` and anime with at least one known producer, to test whether projects with clearly identified producer organizations systematically perform better.  
- Compare the average Score across different collaboration structures based on `num_producers` (for example: 1 producer vs 2 producers vs 3 or more), to quantify whether producer collaboration is associated with higher success than single producer projects.  
- Identify the top 10 producers with the highest average Score (the Score of a co produced anime is fully credited to each producer involved) to highlight which producer brands tend to be associated with successful titles.  
- Restrict to producers that have co produced at least 3 anime and compare their average Score, in order to find producer teams or recurring producer combinations that consistently deliver high performing projects.

#### Deep insight 1

In [23]:
import plotly.express as px
import plotly.io as pio
import pandas as pd
import numpy as np

# 1. Setup Theme
pio.templates.default = "plotly_white"

# --- 2. DATA PREPARATION ---
# Copy dữ liệu
df_compare = df_anime_dataset_2023_prep[['Producers', 'Score']].copy()

# Xử lý cột Score
df_compare['Score'] = pd.to_numeric(df_compare['Score'], errors='coerce')
df_compare = df_compare.dropna(subset=['Score'])

# --- HÀM PHÂN LOẠI (XỬ LÝ CẢ Ô TRỐNG) ---
def categorize_producer(x):
    if pd.isna(x): return 'Unknown'          # Bắt NaN/Null
    if str(x).strip() == "": return 'Unknown' # Bắt chuỗi rỗng
    if str(x).strip().upper() == "UNKNOWN": return 'Unknown' # Bắt chữ Unknown
    return 'Known Producers'

# Áp dụng hàm
df_compare['Producer_Type'] = df_compare['Producers'].apply(categorize_producer)

# Tính toán Mean Insight
mean_scores = df_compare.groupby('Producer_Type')['Score'].mean()
unknown_mean = mean_scores.get('Unknown', 0)
known_mean = mean_scores.get('Known Producers', 0)
diff = known_mean - unknown_mean

# --- 3. VISUALIZATION ---
fig = px.box(
    df_compare,
    x='Producer_Type',
    y='Score',
    color='Producer_Type',
    # Định nghĩa màu: Xám cho Unknown, Tím cho Known
    color_discrete_map={
        'Unknown': '#B0B0B0',       
        'Known Producers': '#8B5CF6' 
    }
)

# --- SỬA LỖI MÀU SẮC TẠI ĐÂY ---
fig.update_traces(
    width=0.4,          # Độ rộng hộp
    marker_size=5,      # Kích thước điểm ngoại lai (outliers)
    marker_opacity=0.4, # Độ mờ điểm ngoại lai
    # marker_color="black"  <-- ĐÃ XÓA DÒNG NÀY ĐỂ KHÔNG BỊ ĐEN HẾT BIỂU ĐỒ
    line_width=2      # Độ dày đường viền hộp
)

# --- 4. LAYOUT CONFIGURATION ---
fig.update_layout(
    title={
        'text': f'<b>Score Comparison: Known vs. Unknown Producers</b><br>'
                f'<span style="font-size: 13px; color: #555;">'
                f'Known Average Score: <b>{known_mean:.2f}</b> vs Unknown Average Score: <b>{unknown_mean:.2f}</b> '
                f'(Difference: <span style="color:green">+{diff:.2f}</span>)</span>',
        'y': 0.95,
        'x': 0.5,
        'xanchor': 'center',
        'yanchor': 'top'
    },
    font=dict(family='Inter', size=13, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    height=800,
    width=1200,
    showlegend=False,
    margin=dict(t=120, l=80, r=50, b=80),

    # --- TRỤC X ---
    xaxis=dict(
        title="",
        tickfont=dict(size=16, weight='bold', color='#333333'),
        showgrid=False,
        showline=True,
        linecolor='#E5E7EB',
        zeroline=False
    ),

    # --- TRỤC Y ---
    yaxis=dict(
        title="Anime Score",
        title_font=dict(size=16, color='#4B5563'),
        tickfont=dict(size=12, color='#6B7280'),
        showgrid=True,
        gridcolor='#F3F4F6',
        showline=True,
        linecolor='#E5E7EB',
        range=[0, 10], 
        dtick=1 
    )
)

fig.show()



**1. Clear Performance Gap (Quantitative Gap)**

  * Average Score (Mean):

    * **Known Producers:** 6.55
    * **Unknown Producers:** 6.11
    * **Difference:** +0.45 points (on a 10-point anime scale, this is a large gap)
  * Median:

    * The purple box (Known) is visibly higher than the grey box (Unknown), meaning most titles with a known producer land in a safer scoring range.

**2. Risk Distribution**

  * **Quality Floor (Worst Cases):**

    * Unknown group shows many low-score outliers (below **3.0–4.0**), suggesting weak funding, low control, or informal production.
  * **Quality Ceiling (Best Cases):**

    * Known group reaches up to **9.0**, showing that major successes (“masterpieces”) almost always come from projects with professional producers involved.

**3. Strategic Understanding (Producer Value)**

  * A producer is not only “money support” but also a **quality filter** providing:

    * Proper production workflow
    * Clear market direction
    * Marketing and distribution resources
  * The “Unknown” label acts as a **warning**, indicating a higher project failure risk and unstable results.

**4. Conclusion**

  * Data shows that having a recognized producer is a critical factor for achieving a stable baseline score and improving the chance of success. Serious projects should not leave this role empty.


#### Deep insight 2

In [24]:
# --- 2. DATA PROCESSING ---
# Copy dữ liệu
df_producers_collab = df_anime_dataset_2023_prep[['Producers', 'Score']].copy()

# Xử lý Score
df_producers_collab['Score'] = pd.to_numeric(df_producers_collab['Score'], errors='coerce')
df_producers_collab = df_producers_collab.dropna(subset=['Score'])

# Loại bỏ các dòng Unknown/Rỗng để việc đếm chính xác
exclude_list = ["UNKNOWN", "", "NONE"]
df_producers_collab = df_producers_collab[~df_producers_collab['Producers'].astype(str).str.strip().str.upper().isin(exclude_list)]
df_producers_collab = df_producers_collab.dropna(subset=['Producers'])

# TÍNH TOÁN SỐ LƯỢNG PRODUCERS
# Đếm số dấu phẩy + 1 (Ví dụ: "A, B" có 1 dấu phẩy -> 2 producers)
df_producers_collab['num_producers'] = df_producers_collab['Producers'].astype(str).str.count(',') + 1

# PHÂN NHÓM (BINNING) ĐỂ BIỂU ĐỒ GỌN GÀNG
def group_size(n):
    if n == 1: return '1 Producer'
    if n == 2: return '2 Producers'
    if n == 3: return '3 Producers'
    if 4 <= n <= 5: return '4-5 Producers'
    return '6+ Producers'

df_producers_collab['Collaboration_Group'] = df_producers_collab['num_producers'].apply(group_size)

# Sắp xếp thứ tự để biểu đồ vẽ từ 1 -> 6+
order_list = ['1 Producer', '2 Producers', '3 Producers', '4-5 Producers', '6+ Producers']

# Tính toán Insight (Mean Score cho từng nhóm)
group_means = df_producers_collab.groupby('Collaboration_Group')['Score'].mean().reindex(order_list)
# Lấy điểm thấp nhất (1 Producer) và cao nhất (thường là nhóm nhiều producers) để so sánh
solo_score = group_means['1 Producer']
max_group_score = group_means.max()
best_group = group_means.idxmax()
diff = max_group_score - solo_score

# --- 3. VISUALIZATION ---
fig = px.box(
    df_producers_collab,
    x='Collaboration_Group',
    y='Score',
    color_discrete_sequence=['#8B5CF6'],
    category_orders={'Collaboration_Group': order_list} # Bắt buộc vẽ đúng thứ tự
)

fig.update_traces(
    width=0.5,
    marker_size=5,
    marker_opacity=0.3,
    line_width=2
)

# --- 4. LAYOUT CONFIGURATION ---
fig.update_layout(
    title={
        'text': f'<b>Impact of Collaboration Size on Anime Score</b><br>'
                f'<span style="font-size: 14px; color: #555;">'
                f'Solo Projects Avg: <b>{solo_score:.2f}</b> vs {best_group} Avg: <b>{max_group_score:.2f}</b> '
                f'(Improvement: <span style="color:green">+{diff:.2f}</span>)</span>',
        'y': 0.95,
        'x': 0.5,
        'xanchor': 'center',
        'yanchor': 'top'
    },
    font=dict(family='Inter', size=14, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    
    width=1200,
    height=800,
    
    showlegend=False, # Ẩn legend vì trục X đã rõ
    margin=dict(t=120, l=100, r=80, b=100),

    # TRỤC X
    xaxis=dict(
        title="Number of Producers Involved",
        title_font=dict(size=16, color='#4B5563'),
        tickfont=dict(size=14, weight='bold', color='#333333'),
        showgrid=False,
        showline=True,
        linecolor='#E5E7EB'
    ),

    # TRỤC Y
    yaxis=dict(
        title="Anime Score",
        title_font=dict(size=16, color='#4B5563'),
        tickfont=dict(size=14, color='#6B7280'),
        showgrid=True,
        gridcolor='#F3F4F6',
        showline=True,
        linecolor='#E5E7EB',
        range=[0, 10],
        dtick=1
    )
)

fig.show()



**Insight: The "Production Committee" Effect (Sức mạnh của Hội đồng Sản xuất)**

Dữ liệu chứng minh rằng mô hình hợp tác (Joint Venture) vượt trội hoàn toàn so với sản xuất đơn lẻ.

*   **Hiệu quả hợp tác (+0.98 điểm):** Có một sự tương quan thuận chiều rõ rệt: số lượng Producer càng tăng, điểm số càng cao. Mức chênh lệch gần **1.0 điểm** giữa dự án "Solo" (6.23) và nhóm "4-5 Producers" (7.21) là minh chứng cho việc **ngân sách lớn hơn và quy trình giám sát chặt chẽ hơn** sẽ tạo ra sản phẩm tốt hơn.
*   **Quản trị rủi ro (Risk Mitigation):** Nhóm **"1 Producer"** có biên độ dao động rất lớn với nhiều "thảm họa" (outliers) rơi xuống mức điểm 2.0 - 3.0. Ngược lại, khi có từ **3 Producers trở lên**, "sàn chất lượng" (the floor) được nâng lên đáng kể (hiếm khi dưới 5.0), giúp bảo vệ khoản đầu tư khỏi thất bại thảm hại.
*   **Điểm tối ưu (The Sweet Spot):** Hiệu suất đạt đỉnh ở nhóm **4-5 Producers**. Nhóm **6+ Producers** dù vẫn tốt nhưng không cho thấy sự tăng trưởng điểm số rõ rệt (bão hòa), gợi ý rằng việc thêm quá nhiều bên tham gia có thể gây ra hiệu suất giảm dần (*diminishing returns*) do khâu quản lý phức tạp.

**👉 Kết luận:** Chiến lược tối ưu là thành lập một Hội đồng sản xuất (Production Committee) gồm **3-5 đối tác** để cân bằng giữa nguồn vốn, kiểm soát chất lượng và hiệu quả quản lý.

In [25]:
import plotly.express as px

# 1. TÍNH TOÁN SỐ LƯỢNG (AGGREGATION)
# Đếm số dòng trong mỗi nhóm
group_counts = df_producers_collab['Collaboration_Group'].value_counts()

# Sắp xếp lại theo đúng thứ tự logic (1 -> 6+) thay vì theo độ lớn
order_list = ['1 Producer', '2 Producers', '3 Producers', '4-5 Producers', '6+ Producers']
group_counts = group_counts.reindex(order_list).reset_index()
group_counts.columns = ['Collaboration_Group', 'Anime_Count']

# Tính thêm phần trăm để hiển thị cho hay
total_anime = group_counts['Anime_Count'].sum()
group_counts['Percentage'] = (group_counts['Anime_Count'] / total_anime * 100).round(1)

# 2. VẼ BIỂU ĐỒ (BAR CHART)
fig = px.bar(
    group_counts,
    x='Collaboration_Group',
    y='Anime_Count',
    text='Anime_Count', # Hiển thị số lượng trên đầu cột
    color_discrete_sequence=['#8B5CF6'] # Màu tím chủ đạo
)

# 3. TINH CHỈNH GIAO DIỆN
fig.update_traces(
    texttemplate='%{text:,}', # Format số có dấu phẩy (ví dụ 1,200)
    textposition='outside',
    marker_line_color='white',
    marker_line_width=1.5,
    # Hover hiển thị cả số lượng và phần trăm
    hovertemplate='<b>%{x}</b><br>Count: %{y:,}<br>Share: %{customdata[0]}%<extra></extra>',
    customdata=group_counts[['Percentage']] # Truyền dữ liệu % vào hover
)

fig.update_layout(
    title={
        'text': f'<b>Distribution of Anime by Collaboration Size</b><br>'
                f'<span style="font-size: 14px; color: #555;">'
                f'Total Analyzed Anime: <b>{total_anime:,}</b></span>',
        'y': 0.95,
        'x': 0.5,
        'xanchor': 'center',
        'yanchor': 'top'
    },
    font=dict(family='Inter', size=14, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    
    width=1200,
    height=700,
    
    showlegend=False,
    margin=dict(t=100, l=80, r=80, b=100),

    # TRỤC X
    xaxis=dict(
        title="Number of Producers Involved",
        title_font=dict(size=16, color='#4B5563'),
        tickfont=dict(size=14, weight='bold', color='#333333'),
        showgrid=False,
        showline=True,
        linecolor='#E5E7EB'
    ),

    # TRỤC Y
    yaxis=dict(
        title="Number of Anime",
        title_font=dict(size=16, color='#4B5563'),
        tickfont=dict(size=14, color='#6B7280'),
        showgrid=True,
        gridcolor='#F3F4F6',
        showline=True,
        linecolor='#E5E7EB',
        tickformat=',' # Format trục Y có dấu phẩy
    )
)

fig.show()

#### Deep Insight 3

In [26]:
# --- 2. DATA PROCESSING ---
# Copy dữ liệu để xử lý
df_quality = df_anime_dataset_2023_prep[['Producers', 'Score']].copy()

# Chuyển đổi Score sang số và loại bỏ dòng không có điểm
df_quality['Score'] = pd.to_numeric(df_quality['Score'], errors='coerce')
df_quality = df_quality.dropna(subset=['Score'])

# --- BƯỚC LÀM SẠCH KỸ LƯỠNG (DEEP CLEANING) ---
df_quality['Producers'] = df_quality['Producers'].astype(str)
# Loại bỏ ký tự thừa [ ] ' "
df_quality['Producers'] = df_quality['Producers'].str.replace(r"[\[\]\'\"]", "", regex=True)
# Tách chuỗi và Explode
df_quality['Producers'] = df_quality['Producers'].str.split(',')
df_quality = df_quality.explode('Producers')
# Cắt khoảng trắng
df_quality['Producers'] = df_quality['Producers'].str.strip()
# Loại bỏ rác
exclude_list = ["UNKNOWN", "NONE", "NAN", "NULL", ""]
df_quality = df_quality[~df_quality['Producers'].str.upper().isin(exclude_list)]

# --- 3. AGGREGATION & FILTERING (QUAN TRỌNG) ---
# Gom nhóm theo Producer và tính 2 chỉ số: Điểm trung bình (mean) và Số lượng phim (count)
producer_stats = df_quality.groupby('Producers')['Score'].agg(['mean', 'count']).reset_index()

# ĐẶT NGƯỠNG LỌC (THRESHOLD)
# Chỉ lấy những nhà sản xuất đã làm ít nhất 10 bộ phim
min_anime_count = 20
top_quality = producer_stats[producer_stats['count'] >= min_anime_count].copy()

# Lấy Top 10 dựa trên điểm trung bình (Mean Score)
top_quality = top_quality.sort_values('mean', ascending=False).head(20)

# Sắp xếp lại Tăng dần để vẽ biểu đồ ngang (người điểm cao nhất nằm trên cùng)
top_quality = top_quality.sort_values('mean', ascending=True)

# --- 4. VISUALIZATION ---
fig = px.bar(
    top_quality,
    x='mean',
    y='Producers',
    orientation='h',
    text='mean', # Hiển thị điểm số trên thanh
    
    title=f'<b>Top 10 High-Quality Producers</b> (Min. {min_anime_count} Titles)',
    
    # Tô màu theo điểm số (Càng cao càng đậm)
    color='mean',
    color_continuous_scale=['#DDD6FE', '#4C1D95']
)

# Tinh chỉnh Traces
fig.update_traces(
    texttemplate='%{text:.2f}', # Format điểm số lấy 2 số thập phân
    textposition='outside',
    marker_line_color='white',
    marker_line_width=1.5,
    # Hover hiển thị đầy đủ: Tên, Điểm, và Số lượng phim đã làm
    hovertemplate='<b>%{y}</b><br>' +
                  'Avg Score: <b>%{x:.2f}</b><br>' +
                  'Total Anime: %{customdata[0]}<extra></extra>',
    customdata=top_quality[['count']] # Truyền dữ liệu số lượng vào hover
)

# --- 5. LAYOUT CONFIGURATION ---
fig.update_layout(
    title={
        'text': f'<b>Top 20 High-Quality Producers (Quality over Quantity)</b><br>'
                f'<span style="font-size: 14px; color: #555;">'
                f'Criteria: Producers with at least {min_anime_count} releases</span>',
        'y': 0.95,
        'x': 0.5,
        'xanchor': 'center',
        'yanchor': 'top'
    },
    font=dict(family='Inter', size=14, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    
    width=1200,
    height=700,
    
    showlegend=False,
    coloraxis_showscale=False, # Ẩn thanh màu bên phải
    margin=dict(t=100, l=250, r=80, b=80), # Margin trái lớn cho tên Producer

    # TRỤC X: Điểm số
    xaxis=dict(
        title="Average Anime Score",
        title_font=dict(size=16, color='#4B5563'),
        tickfont=dict(size=14, weight='bold', color='#333333'),
        showgrid=True,
        gridcolor='#F3F4F6',
        showline=True,
        linecolor='#E5E7EB',
        range=[0, 10], # Giữ thang điểm 10 để trung thực
        dtick=1
    ),

    # TRỤC Y: Tên Producer
    yaxis=dict(
        title="",
        tickfont=dict(size=14, color='#4B5563'),
        ticksuffix="  "
    )
)

fig.show()


 **DEEP INSIGHT: CÔNG THỨC "CHẤT LƯỢNG" CỦA NHỮNG NHÀ SẢN XUẤT HÀNG ĐẦU**

Từ danh sách Top 20 Producer có điểm số trung bình cao nhất (Criteria > 20 releases)

**1. Sức mạnh tối thượng của "IP Holding" (Sở hữu trí tuệ gốc)**

*   **Dẫn chứng:** Vị trí **Top 1 thuộc về Notes (7.65)**.
*   **Phân tích:** *Notes* chính là công ty quản lý thương hiệu **Type-Moon** (nổi tiếng với *Fate series, Kara no Kyoukai*).
    *   Họ không sản xuất đại trà. Họ là chủ sở hữu IP gốc. Khi họ đứng ra làm Producer, đó là để bảo vệ và nâng tầm thương hiệu "con cưng" của họ.
    *   **Bài học:** Khi Producer chính là "cha đẻ" của tác phẩm (hoặc nắm quyền kiểm soát tuyệt đối IP), họ sẽ chấp nhận chi ngân sách lớn (High Budget) và kiểm soát chất lượng cực gắt gao (Strict QC). Điều này dẫn đến điểm số cao vượt trội và lượng fan trung thành tuyệt đối.

**2. Chiến lược "Cherry-picking" của các Nhà xuất bản (Publishers)**
*   **Dẫn chứng:** Sự xuất hiện dày đặc của các ông lớn xuất bản Manga: **Hakusensha (#3), Kodansha (#10), Houbunsha (#12), Shueisha (#14)**.
*   **Phân tích:** Tại sao các nhà xuất bản lại có điểm cao?
    *   Họ sở hữu hàng nghìn đầu truyện Manga, nhưng họ chỉ chọn lọc (cherry-pick) những bộ **bán chạy nhất, nội dung hay nhất** để chuyển thể thành Anime.
    *   Nội dung gốc đã được thị trường kiểm chứng (Validated Content) $\rightarrow$ Rủi ro kịch bản thấp $\rightarrow$ Anime dễ dàng đạt điểm cao.
    *   *Đối chiếu:* Họ làm ít phim hơn các đài truyền hình (như NHK, TV Tokyo ở BXH Số lượng), nhưng tỷ lệ thành công (Hit-rate) cao hơn hẳn.

**3. " Quy mô lớn nhưng Chất lượng vẫn cao**
*   **Dẫn chứng:** 
    * Aniplex - Top 20 Active Producers (580 projects) & Top 20 High-Quality Producers (7.34 avg score)
    * Kodansha - Top 20 Active Producers (310 projects) & Top 20 High-Quality Producers (7.43 avg score)
    * Dentsu - Top 20 Active Producers (370 projects) & Top 20 High-Quality Producers (7.41 avg score)
    * Shueisha - Top 20 Active Producers (307 projects) & Top 20 High-Quality Producers (7.39 avg score)
    * Mainichi Broadcasting System - Top 20 Active Producers (262 projects) & Top 20 High-Quality Producers (7.38 avg score)
*   **Phân tích chéo (Cross-Reference):**
    * Trong khi các "trùm số lượng" khác như *NHK, TV Tokyo, Lantis* đều không xuất hiện trong bảng xếp hạng chất lượng, thì 5 cái tên nổi bật trên vẫn trụ lại trong bảng xếp hạng chất lượng.



#### Deep Insight 4


In [45]:


# --- 2. DATA PROCESSING ---
# Copy dữ liệu
df_producers_quality_collab = df_anime_dataset_2023_prep[['Producers', 'Score']].copy()

# Xử lý Score
df_producers_quality_collab['Score'] = pd.to_numeric(df_producers_quality_collab['Score'], errors='coerce')
df_producers_quality_collab = df_producers_quality_collab.dropna(subset=['Score'])

# --- LÀM SẠCH CHUỖI NHƯNG KHÔNG TÁCH (KEEP COMBINATIONS) ---
# Chỉ loại bỏ các ký tự rác, giữ nguyên dấu phẩy để nhận diện tổ hợp
df_producers_quality_collab['Producers'] = df_producers_quality_collab['Producers'].astype(str).str.replace(r"[\[\]\'\"]", "", regex=True)
df_producers_quality_collab['Producers'] = df_producers_quality_collab['Producers'].str.strip()

# Loại bỏ các dòng Unknown/Rỗng
exclude_list = ["UNKNOWN", "NONE", "NAN", "NULL", ""]
df_producers_quality_collab = df_producers_quality_collab[~df_producers_quality_collab['Producers'].str.upper().isin(exclude_list)]

# (Tùy chọn) Chỉ lấy các dòng có ít nhất 1 dấu phẩy (tức là có sự hợp tác của >= 2 bên)
# Nếu bạn muốn tính cả các Studio đơn lẻ thì bỏ dòng dưới đi. 
# Nhưng "Synergy" (Hợp tác) thường ám chỉ 2 bên trở lên.
df_producers_quality_collab = df_producers_quality_collab[df_producers_quality_collab['Producers'].str.contains(',')]

# --- 3. AGGREGATION ---
# Gom nhóm theo đúng chuỗi Combination
combo_stats = df_producers_quality_collab.groupby('Producers')['Score'].agg(['mean', 'count']).reset_index()

# --- FILTERING (CRITERIA) ---
# 1. Phải hợp tác ít nhất 3 lần (Độ ổn định)
min_collabs = 5
top_combos = combo_stats[combo_stats['count'] >= min_collabs].copy()

# 2. Lấy Top 15 cặp đôi có điểm cao nhất
top_combos = top_combos.sort_values('mean', ascending=False).head(20)

# Sắp xếp tăng dần để vẽ biểu đồ ngang
top_combos = top_combos.sort_values('mean', ascending=True)

# --- 4. VISUALIZATION ---
fig = px.bar(
    top_combos,
    x='mean',
    y='Producers',
    orientation='h',
    text='mean',
    
    title=f'<b>Top 15 "Dream Teams": Best Producer Collaborations</b>',
    
    # Tô màu theo điểm số
    color='mean',
    color_continuous_scale=['#DDD6FE', '#4C1D95']
)

# --- 5. TINH CHỈNH GIAO DIỆN ---
fig.update_traces(
    texttemplate='%{text:.2f}', 
    textposition='outside',
    marker_line_color='white',
    marker_line_width=1.5,
    # Hover hiển thị: Tên tổ hợp, Điểm, Số lần hợp tác
    hovertemplate='<b>Combination:</b> %{y}<br>' +
                  'Avg Score: <b>%{x:.2f}</b><br>' +
                  'Collabs Count: %{customdata[0]}<extra></extra>',
    customdata=top_combos[['count']]
)

fig.update_layout(
    title={
        'text': f'<b>Top 15 Best Collaborations Producer Teams (Quality Focus)</b><br>'
                f'<span style="font-size: 14px; color: #555;">'
                f'Criteria: Combinations with ≥ {min_collabs} joint projects</span>',
        'y': 0.95, 'x': 0.5, 'xanchor': 'center', 'yanchor': 'top'
    },
    font=dict(family='Inter', size=13, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    
    width=1400,
    height=800,
    
    showlegend=False,
    coloraxis_showscale=False,
    
    # Margin trái RẤT LỚN vì tên tổ hợp thường rất dài
    margin=dict(t=100, l=400, r=60, b=80),

    # TRỤC X
    xaxis=dict(
        title="Average Anime Score",
        title_font=dict(size=14, color='#4B5563'),
        tickfont=dict(size=12, color='#333333'),
        showgrid=True,
        gridcolor='#F3F4F6',
        zeroline=False,
        # Zoom vào vùng điểm cao để thấy sự chênh lệch
        range=[top_combos['mean'].min() - 0.5, 10] 
    ),

    # TRỤC Y
    yaxis=dict(
        title="",
        tickfont=dict(size=12, color='#4B5563'),
        ticksuffix="  "
    )
)

fig.show()

<!-- **Deep Insight**:
- Hãy cho biết điểm trung bình score giữa việc colab và tự thực hiện 1 mình thì cái nào hiệu quả hơn.
- Hãy cho biết top 10 Producer đạt được điểm trung bình Score cao nhất (điểm score của anime khi colab cũng được tính là điểm riêng cho producer tham colab)
- Đội Producers nào có điểm score trung bình cao nhất (chỉ tính các đội colab trên 3 movie anime) -->

## 3.8.Insight Question 8
Unlike Producers, Studios are hired by Producers to directly create the anime. They are responsible for drawing, animation, compositing, coloring, editing, and post production. In other words, they are the teams that produce the actual visual content you see on the screen. Given that animation production can be a massive undertaking, it's common for multiple studios to collaborate on a single anime, especially for larger projects or to meet tight deadlines. As a result, the 'Studios' feature in our dataset often lists several animation studios within a single entry, separated by commas. Our initial exploration of this raw column will aim to identify the most frequently appearing combinations of studios. By analyzing this unprocessed, aggregated data, we intend to highlight how its current string format will impede the identification of individual studios' prevalence and necessitate significant parsing and cleaning to understand the true contributors to anime production.

### 3.8.1.Misleading Insight
- What are the most frequently appearing studio combinations or studio strings as they are originally listed in the dataset?

In [58]:


# --- 2. DATA PROCESSING ---
df_studios = df_anime_dataset_2023_prep[['Studios']].copy()

# Chuyển về dạng string
df_studios['Studios'] = df_studios['Studios'].astype(str)

# Loại bỏ các ký tự rác (ngoặc, nháy) để đếm cho chuẩn
df_studios['Studios'] = df_studios['Studios'].str.replace(r"[\[\]\'\"]", "", regex=True)

# Lọc bỏ Unknown/None
exclude_list = ["UNKNOWN", "NONE", "NAN", "NULL", ""]
df_studios = df_studios[~df_studios['Studios'].str.upper().isin(exclude_list)]

# --- LOGIC PHÂN LOẠI (SOLO VS COLLAB) ---
def classify_studio(studio_str):
    # Đếm số lượng phần tử sau khi tách dấu phẩy
    # Ví dụ: "Madhouse" -> 1 phần tử -> Solo
    # Ví dụ: "Madhouse, MAPPA" -> 2 phần tử -> Collab
    count = len(studio_str.split(','))
    if count == 1:
        return 'Solo Studio'
    else:
        return 'Collaboration'

df_studios['Production_Type'] = df_studios['Studios'].apply(classify_studio)

# --- 3. AGGREGATION ---
studio_type_counts = df_studios['Production_Type'].value_counts().reset_index()
studio_type_counts.columns = ['Type', 'Count']

# Tính phần trăm thủ công để đưa vào title (Insight nhanh)
total = studio_type_counts['Count'].sum()
solo_count = studio_type_counts[studio_type_counts['Type'] == 'Solo Studio']['Count'].values[0]
collab_count = studio_type_counts[studio_type_counts['Type'] == 'Collaboration']['Count'].values[0]
solo_pct = (solo_count / total) * 100

# --- 4. VISUALIZATION (DONUT CHART) ---
fig = px.pie(
    studio_type_counts,
    values='Count',
    names='Type',
    hole=0.5, # Tạo lỗ hổng ở giữa (Donut Chart)
    
    # Màu sắc: Tím đậm cho Collab (tượng trưng sự phức tạp), Tím nhạt cho Solo
    color='Type',
    color_discrete_map={
        'Solo Studio': '#C4B5FD',    # Tím nhạt
        'Collaboration': '#6D28D9'   # Tím đậm
    },
    
    title='<b>Production Structure: Solo vs. Collaboration Studios</b>'
)

# --- 5. TINH CHỈNH GIAO DIỆN ---
fig.update_traces(
    textposition='outside', 
    textinfo='percent+label', # Hiển thị Nhãn + %
    textfont=dict(size=13, family='Inter', color='#333333'),
    marker=dict(line=dict(color='white', width=3)) # Viền trắng ngăn cách
)

fig.update_layout(
    font=dict(family='Inter', size=14, color="#333333"),
    
    # Thêm Annotation ở giữa lỗ tròn để nhấn mạnh con số tổng
    annotations=[dict(
        text=f'<b>{total:,}</b><br><span style="font-size:12px; color:gray">Total Projects</span>', 
        x=0.5, y=0.5, font_size=20, showarrow=False
    )],
    
    title={
        'text': f'<b>Studio Production Structure</b><br>'
                f'<span style="font-size: 14px; color: #555;">'
                f'Solo Projects dominate with <b>{solo_pct:.1f}%</b> of the market</span>',
        'y': 0.95, 'x': 0.5, 'xanchor': 'center'
    },
    
    showlegend=False, # Ẩn chú thích bên cạnh vì đã có nhãn trực tiếp
    height=600,
    width=800,
    margin=dict(t=100, b=50, l=50, r=50)
)

fig.show()

In [56]:

# 1. Xử lý dữ liệu (Top 15 Studios - Raw Data)
# Đếm tần suất xuất hiện của các chuỗi trong cột Studios
studio_counts = df_anime_dataset_2023['Studios'].value_counts().head(10).reset_index()
studio_counts.columns = ['Studios', 'Count']

# Sắp xếp ascending=True để khi vẽ ngang, thanh lớn nhất (Top 1) nằm trên cùng
studio_counts = studio_counts.sort_values('Count', ascending=True)

# 2. Logic màu sắc (Gray cho Unknown, Tím cho còn lại)
base_color = "#8B5CF6"
unknown_color = "#B0B0B0"

# Tạo list màu dựa trên tên Studios
colors = [
    unknown_color if str(s).strip().upper() == "UNKNOWN" else base_color
    for s in studio_counts['Studios']
]

# 3. Tạo biểu đồ Horizontal Bar
fig = px.bar(
    studio_counts,
    x='Count',       # Số lượng nằm ngang
    y='Studios',     # Tên nằm dọc
    orientation='h', # Xoay ngang
    title='Top 15 Studio Combinations (Raw Data)',
    labels={'Studios': 'Studio Combination', 'Count': 'Number of Anime'},
    text='Count'
)

# 4. Tinh chỉnh traces
fig.update_traces(
    marker_color=colors,  # Áp dụng list màu đã tạo
    marker_line_color='white',
    marker_line_width=1.2,
    textposition='outside',
    texttemplate='%{text:,}',
    hovertemplate='<b>Studios:</b> %{y}<br><b>Count:</b> %{x:,}<extra></extra>'
)

# 5. Tinh chỉnh Layout (Style đồng bộ)
fig.update_layout(
    title={
        'text': 'Top 15 Studio Combinations (Raw Data)',
        'x': 0.5,
        'xanchor': 'center',
        'font': dict(size=22, family='Inter', weight='bold', color='#1F2937')
    },
    # Trục X: Số lượng
    xaxis=dict(
        title='Number of Anime',
        title_font=dict(size=14, color='#4B5563', family='Inter'),
        tickfont=dict(size=11, color='#6B7280'),
        showgrid=False,       # Tắt grid theo code mẫu của bạn
        gridcolor='#F3F4F6',
        showline=False,
        zeroline=False,
        tickformat=','
    ),
    # Trục Y: Tên Studios
    yaxis=dict(
        title='',  # Bỏ tiêu đề trục
        tickfont=dict(size=14, color='#4B5563', family='Inter'),
        ticksuffix="  ",  # Tạo khoảng cách giữa chữ và thanh
        showgrid=False,
        showline=False,
        zeroline=False
    ),
    plot_bgcolor='white',
    paper_bgcolor='white',
    bargap=0.25,
    height=800,
    
    # Margin left (l) để 250 là an toàn cho tên Studio (thường ngắn hơn Producer chút nhưng vẫn dài)
    margin=dict(t=80, l=250, r=40, b=80),
    
    hoverlabel=dict(
        bgcolor='white',
        bordercolor='#8B5CF6',
        font=dict(size=13, family='Inter', color='#1F2937')
    )
)

fig.show()

**Conflict**

The initial insights derived from the **'Studios'** feature (showing the frequency of studio combinations) are significantly flawed and misleading due to critical data integrity and structural issues. The raw state of this column prevents an accurate understanding of individual studios' contributions or true prevalence.

*   **Error 1: Aggregated Studio Strings (Multiple Studios in One Entry).**
    *   **Impact on Insights (Bar Charts of Studio Combinations):** The primary issue is that the **'Studios'** column often contains multiple studios listed within a single string entry (e.g., **"Madhouse, MAPPA"**). When we count the frequency of these strings, we are tallying specific combinations of studios, not the individual studios themselves. This fundamentally obscures the true involvement and popularity of any single animation studio. For example, **"Madhouse"** might be a highly active studio, but its individual prevalence is hidden within numerous unique combinations. The bar charts will show these distinct, often lengthy, combination strings as categories, making it impossible to determine which individual studios are most active or involved across the dataset. We are gaining insight into **studio collaborations**, rather than the independent entities.

*   **Error 2: Non-standard Missing Value Representation - 'UNKNOWN' instead of NaN.**
    *   **Impact on Insights (Bar Charts of Studio Combinations):** The presence of the string 'UNKNOWN' for missing studio information is treated as a valid and distinct studio combination by Plotly. This means 'UNKNOWN' will likely appear as one of the most frequent "studio combinations" in the bar charts. Its inclusion as a prominent category inflates its perceived importance, distorts the true ranking of actual **studio collaborations**, and distracts from analyzing meaningful production data. A correctly handled missing value (standard NaN) would typically be excluded from such frequency counts by default, providing a clearer and more accurate insight into actual studio involvement.

### 3.8.2.Correct Insight

In [72]:


# --- 2. DATA PROCESSING (DEEP CLEANING FOR STUDIOS) ---
# Lấy cột Studios thay vì Producers
studio_df = df_anime_dataset_2023_prep[['Studios']].copy()
studio_df['Studios'] = studio_df['Studios'].astype(str)

# REGEX CLEANING
studio_df['Studios'] = studio_df['Studios'].str.replace(r"[\[\]\'\"]", "", regex=True)

# SPLIT & EXPLODE (Tách các Studio hợp tác ra đếm riêng)
studio_df['Studios'] = studio_df['Studios'].str.split(',')
studio_df = studio_df.explode('Studios')

# TRIM
studio_df['Studios'] = studio_df['Studios'].str.strip()

# FILTER GARBAGE
exclude_list = ["UNKNOWN", "NONE", "NAN", "NULL"]
studio_df = studio_df[~studio_df['Studios'].str.upper().isin(exclude_list)]
studio_df = studio_df[studio_df['Studios'] != ""]

# --- 3. AGGREGATION ---
# Đếm số lượng và lấy Top 20
top_studios = studio_df['Studios'].value_counts().head(25).reset_index()
top_studios.columns = ['Studios', 'Count']
top_studios = top_studios.sort_values('Count', ascending=True)

# --- 4. VISUALIZATION ---
fig = px.bar(
    top_studios,
    x='Count',
    y='Studios',
    orientation='h',
    text='Count',
    title='<b>Market Leaders: Top 25 Most Active Individual Studios</b>',
    color='Count',
    # Giữ nguyên dải màu Tím
    color_continuous_scale=['#DDD6FE', '#4C1D95'] 
)

# Tinh chỉnh Traces
fig.update_traces(
    texttemplate='%{text:,}',
    textposition='outside',
    hovertemplate='<b>Studio:</b> %{y}<br><b>Projects:</b> %{x:,}<extra></extra>'
)

# --- 5. LAYOUT ---
fig.update_layout(
    title={
        'text': '<b>Top 25 Most Active Individual Studios</b>',
        'y': 0.95,
        'x': 0.5,
        'xanchor': 'center',
        'yanchor': 'top'
    },
    font=dict(family='Inter', size=13, color="#333333"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    height=750, 
    coloraxis_showscale=False,
    margin=dict(t=80, l=150, r=50, b=80),

    # --- CẤU HÌNH TRỤC X (Số lượng) ---
    xaxis=dict(
        title="Number of Anime Produced",
        title_font=dict(size=14, color='#4B5563', family='Inter'),
        showticklabels=True, 
        tickfont=dict(size=14, color='#6B7280'),
        
        # Grid settings
        showgrid=False, # Nên bật Grid mờ để dễ dóng hàng
        gridcolor='#F3F4F6',
        zeroline=False,
        
        # --- QUAN TRỌNG: ĐIỀU CHỈNH BƯỚC NHẢY CHO STUDIO ---
        tick0=0,     
        dtick=200,  
        tickformat=',' 
    ),

    # --- CẤU HÌNH TRỤC Y (Tên Studios) ---
    yaxis=dict(
        title="",
        tickfont=dict(size=14, color='#333333', family='Inter'),
        ticksuffix="  ",    # Khoảng cách đẹp
        showgrid=False,
        showline=False,
        zeroline=False
    )
)

fig.show()

In [ ]:

# Convert 'Genres' column from string representation of list to actual list
# Use errors='coerce' to turn unparseable strings into NaN
df_anime_dataset_2023_prep['Genres_list'] = df_anime_dataset_2023_prep['Genres'].apply(lambda x: ast.literal_eval(x) if isinstance(x, str) else x)
genre_score_pairs = []

# Iterate through each row of the DataFrame
for index, row in df_anime_dataset_2023_prep.iterrows():
    genres = row['Genres_list']
    score = row['Score']

    # Check if both genres and score are not null
    if isinstance(genres, list) and not pd.isna(score):
        for genre in genres:
            genre_score_pairs.append({'Genre': genre, 'Score': score})

# Convert the list of dictionaries to a DataFrame
genre_scores_df = pd.DataFrame(genre_score_pairs)

# Group by genre and calculate the mean score
average_score_per_genre = genre_scores_df.groupby('Genre')['Score'].mean().reset_index()

# Sort by average score in descending order
average_score_per_genre = average_score_per_genre.sort_values(by='Score', ascending=False)

print("Top 10 Genres by Average Score:")
display(average_score_per_genre.head(10))

Top 10 Genres by Average Score:


Genre     Score
3   Award Winning  7.296308
14        Mystery  6.995093
20       Suspense  6.962963
6           Drama  6.850645
15        Romance  6.804509
19   Supernatural  6.744600
18         Sports  6.722046
0          Action  6.674112
1       Adventure  6.673997
11        Gourmet  6.627664

<!-- **Conflict**: Có thể là chỉ có lỗi do mising value

*   List item
*   List item
 -->


In [67]:
i
# --- 1. DATA PREPARATION ---

# Hàm parse list an toàn
def parse_list_field(x):
    if pd.isna(x): return []
    if isinstance(x, list): return [str(s).strip() for s in x]
    try:
        val = ast.literal_eval(x)
        if isinstance(val, (list, tuple)): return [str(s).strip() for s in val]
    except Exception: pass
    return [s.strip() for s in str(x).split(',') if s.strip()]

# Tạo dataframe làm việc
df_matrix = df_anime_dataset_2023_prep[['anime_id', 'Studios', 'Genres', 'Score']].copy()
df_matrix = df_matrix.dropna(subset=['Studios', 'Genres', 'Score'])

# Parse columns
df_matrix['Studios_list'] = df_matrix['Studios'].apply(parse_list_field)
df_matrix['Genres_list'] = df_matrix['Genres'].apply(parse_list_field)

# --- BƯỚC 1: LỌC TOP 20 STUDIOS (Market Leaders) ---
df_studios = df_matrix.explode('Studios_list')
df_studios['Studio'] = df_studios['Studios_list'].astype(str).str.strip()

top_20_studio_names = df_studios['Studio'].value_counts().head(20).index.tolist()
df_top_studios = df_studios[df_studios['Studio'].isin(top_20_studio_names)].copy()

# --- BƯỚC 2: LỌC TOP 10 GENRES (Major Markets) ---
df_final = df_top_studios.explode('Genres_list')
df_final['Genre'] = df_final['Genres_list'].astype(str).str.strip()

# Chỉ lấy 10 Genre xuất hiện nhiều nhất trong tập dữ liệu của Top 20 Studio
top_10_genres = df_final['Genre'].value_counts().head(10).index.tolist()
df_final = df_final[df_final['Genre'].isin(top_10_genres)]

# --- BƯỚC 3: TẠO MA TRẬN ĐIỂM SỐ (PIVOT TABLE) ---
heatmap_data = df_final.groupby(['Studio', 'Genre'])['Score'].mean().unstack()

# Reindex để sắp xếp gọn gàng
heatmap_data = heatmap_data.reindex(index=top_20_studio_names, columns=top_10_genres)

# --- 2. VISUALIZATION (HEATMAP) ---

# Giữ nguyên dải màu trong code gốc của bạn
custom_colorscale = [
    [0.0, "#F3E5F5"], 
    [0.6, "#9575CD"], 
    [1.0, "#4A235A"]  
]

fig = px.imshow(
    heatmap_data,
    # [EDIT 1] Đổi label Avg Score -> Average Score
    labels=dict(x="Genre", y="Studio", color="Average Score"),
    x=heatmap_data.columns,
    y=heatmap_data.index,
    aspect="auto",
    color_continuous_scale=custom_colorscale,
    text_auto=".2f",
    title="<b>Top 20 Active Individual Studios Specialization: Performance in Top 10 Genres</b>"
)

# --- 3. STYLING ---
fig.update_layout(
    height=800, 
    width=1000,
    
    font=dict(family='Roboto', size=11, color="#2c3e50"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    
    # [EDIT 2] Tinh chỉnh Title để chữ Genre không bị dính
    title=dict(
        text="<b>Top 20 Active Individual Studios Specialization: Performance in Top 10 Genres</b>",
        y=0.95,
        x=0.5,
        xanchor='center',
        yanchor='top',
        pad=dict(b=30) # Đẩy nội dung bên dưới (trục Genre) ra xa Title 30px
    ),
    
    # Tăng margin Top lên để chứa phần padding mới
    margin=dict(t=100, l=150, r=50, b=50),
    
    coloraxis_colorbar=dict(
        title="Average Score", # [EDIT 1] Đổi title colorbar
        title_font=dict(family="Inter", size=11),
        tickfont=dict(family="Inter", size=10),
        thickness=10, 
        len=0.5,
        outlinecolor="white", 
        outlinewidth=0
    )
)

# Trục X lên đầu (Matrix Style)
fig.update_xaxes(side="top", tickfont=dict(family="Inter", size=11))
fig.update_yaxes(tickfont=dict(family="Inter", size=11), title=None, ticksuffix="  ")

# Tạo khoảng cách giữa các ô
fig.update_traces(xgap=1, ygap=1)

fig.show()

#### Deep insight 1
**1. Kyoto Animation (KyoAni): Studio toàn diện nhất**
*   **Quan sát:** Dòng cuối cùng của KyoAni là dòng có màu **tím đậm nhất** và đồng đều nhất trên toàn biểu đồ, đặc biệt là các thể loại:
    *   **Action (7.98), Drama (7.98), Mystery (7.82)**.
*   **Insight Chiến lược:** KyoAni không chỉ là "trùm" của các thể loại *Slice of Life* hay *Drama* truyền thống mà còn là Studio có khả năng thực thi các thể loại phức tạp như *Action* hay *Mystery* ở cấp độ cao nhất (**~8.00** điểm).
    *   **Quyết định:** Nếu mục tiêu là **tối đa hóa điểm số** và **uy tín thương hiệu** (Brand Prestige) với ngân sách không giới hạn, KyoAni là lựa chọn ưu tiên số một, vượt qua cả Production I.G và Madhouse.
*   Bên cạnh đó **A-1 Pictures** cũng cho thấy sự toàn diện của mình trong tất cả các lĩnh vực khi điểm trung bình đều trên 7. Tuy nhiên điều đó lại khiến studio này không thực sự nổi bật trong thể loại nào 

 **2. Phân Cấp Studio theo Thể loại (Tiering by Genre)**
Các Studio lớn khác cho thấy sự chuyên môn hóa rõ rệt:

| Studio | Thể loại Mạnh nhất (Tím Đậm) |
| :--- | :--- | 
| **Bones** | Action (**7.42**), Fantasy (**7.53**) | 
| **Shaft** | Mystery (**8.07**), Supernatural (**7.60**) | 
| **TMS Entertainment** | Supernatural (**7.57**), Romance (**7.46**), Drama (**7.40**) |
| **Production I.G** | Sports (**7.59**), Mystery (**7.40**)
| **Madhouse** | Sports (**7.57**) | 

*   **Actionable Insight:**
    *   Producer muốn làm **Action** nên ưu tiên **Bones** (chuyên biệt), **KyoAni** (chất lượng tuyệt đối), hoặc **Madhouse**.
    *   Producer muốn làm **Tâm lý/Tình cảm** nên chọn **TMS Entertainment** hoặc **KyoAni**.
    *   **Cạm bẫy:** Tránh giao dự án *Slice of Life* cho **Bones** và *Sci-Fi/Fantasy* cho **Pierrot** (điểm thấp 6.51 - 7.13) vì không phải sở trường.

 **3. Sàn Chất Lượng (The Baseline Risk)**
*   **Quan sát:** Các Studio như **Tatsunoko Production, AIC, OLM** có rất nhiều ô màu tím nhạt (điểm trung bình quanh 6.50 - 6.80) hoặc có ô màu trắng (dữ liệu thiếu/chưa đủ dự án).
*   **Insight:** Đây là những Studio an toàn cho các dự án **ngân sách trung bình** hoặc **đầu tư mạo hiểm**, nhưng chúng ta không nên kỳ vọng chúng sẽ tạo ra siêu phẩm. Điểm yếu của họ là thiếu sự nhất quán.



In [69]:

# --- 1. DATA PREPARATION ---

# Hàm parse list an toàn (Giữ nguyên)
def parse_list_field(x):
    if pd.isna(x): return []
    if isinstance(x, list): return [str(s).strip() for s in x]
    try:
        val = ast.literal_eval(x)
        if isinstance(val, (list, tuple)): return [str(s).strip() for s in val]
    except Exception: pass
    return [s.strip() for s in str(x).split(',') if s.strip()]

# Tạo dataframe làm việc
df_matrix = df_anime_dataset_2023_prep[['anime_id', 'Studios', 'Type', 'Score']].copy()
df_matrix['Score'] = pd.to_numeric(df_matrix['Score'], errors='coerce')
df_matrix = df_matrix.dropna(subset=['Studios', 'Type', 'Score'])

# Parse Studios list
df_matrix['Studios_list'] = df_matrix['Studios'].apply(parse_list_field)

# Lọc các list rỗng/UNKNOWN/NONE sau khi parse
def filter_empty_or_unknown(lst):
    cleaned_list = [s.strip().upper() for s in lst if s.strip()]
    return len(cleaned_list) > 0 and 'UNKNOWN' not in cleaned_list and 'NONE' not in cleaned_list

df_matrix = df_matrix[df_matrix['Studios_list'].apply(filter_empty_or_unknown)]


# --- BƯỚC 1: LỌC TOP 20 STUDIOS (Market Leaders) ---
df_studios = df_matrix.explode('Studios_list')
df_studios['Studio'] = df_studios['Studios_list'].astype(str).str.strip()

# Lấy Top 20 Studio theo số lượng Anime đã làm
top_20_studio_names = df_studios['Studio'].value_counts().head(20).index.tolist()
df_final = df_studios[df_studios['Studio'].isin(top_20_studio_names)].copy()


# --- BƯỚC 2: TẠO MA TRẬN ĐIỂM SỐ (PIVOT TABLE) ---
# TÍNH ĐIỂM TRUNG BÌNH CỦA TỪNG CẶP STUDIO-TYPE
heatmap_data = df_final.groupby(['Studio', 'Type'])['Score'].mean().unstack()

# Sắp xếp lại thứ tự cột theo logic (TV, OVA, Movie...)
type_order = ['TV', 'Movie', 'OVA', 'Special', 'ONA', 'Music']
# Lấy các cột Type có tồn tại trong dữ liệu
type_columns = [t for t in type_order if t in heatmap_data.columns] 

# Reindex để sắp xếp gọn gàng
heatmap_data = heatmap_data.reindex(index=top_20_studio_names, columns=type_columns)

# --- 2. VISUALIZATION (HEATMAP) ---
custom_colorscale = [
    [0.0, "#F3E5F5"], 
    [0.6, "#9575CD"], 
    [1.0, "#4A235A"]  
]

fig = px.imshow(
    heatmap_data,
    labels=dict(x="Anime Type", y="Studio", color="Average Score"),
    x=heatmap_data.columns,
    y=heatmap_data.index,
    aspect="auto", # Tỉ lệ sẽ hợp lý hơn vì số lượng Type ít
    color_continuous_scale=custom_colorscale,
    text_auto=".2f",
    title="<b>Studio Specialization: Performance by Anime Type</b>"
)

# --- 3. STYLING ---
fig.update_layout(
    height=800, 
    width=1000, 
    
    font=dict(family='Inter', size=12, color="#2c3e50"),
    plot_bgcolor='white',
    paper_bgcolor='white',
    
    title=dict(
        text="<b>Top 20 Studios Specialization: Performance by Anime Type</b>",
        y=0.95,
        x=0.5,
        xanchor='center',
        yanchor='top',
        pad=dict(b=30)
    ),
    
    margin=dict(t=100, l=150, r=50, b=50),
    
    coloraxis_colorbar=dict(
        title="Average Score", 
        title_font=dict(family="Inter", size=11),
        tickfont=dict(family="Inter", size=10),
        thickness=10, 
        len=0.5,
        outlinecolor="white", 
        outlinewidth=0
    )
)

# Trục X lên đầu (Matrix Style)
fig.update_xaxes(
    side="top", 
    tickfont=dict(family="Inter", size=13), 
    tickangle=0, # Không cần xoay chữ vì tên Type ngắn
    title="Anime Type"
)

fig.update_yaxes(
    tickfont=dict(family="Inter", size=11), 
    title="Studio",
    ticksuffix="  "
)

# Tạo khoảng cách giữa các ô
fig.update_traces(xgap=1, ygap=1)

fig.show()

# 4.Deep Insight